src/vivarium/framework/results/context.py

```diff
@@ -79,57 +79,78 @@ def add_observation(
         self,
         name: str,
         pop_filter: str,
+        aggregator_sources: List[str],
         aggregator: Callable[[pd.DataFrame], float],
         additional_stratifications: List[str] = (),
         excluded_stratifications: List[str] = (),
         when: str = "collect_metrics",
         **additional_keys: str,
     ):
         self._warn_check_stratifications(additional_stratifications, excluded_stratifications)
-        groupers = self._get_groupers(additional_stratifications, excluded_stratifications)
-        self._observations[when][(pop_filter, groupers)].append(
-            (name, aggregator, additional_keys)
+        stratifications = self._get_stratifications(
+            additional_stratifications, excluded_stratifications
+        )
+        self._observations[when][(pop_filter, stratifications)].append(
+            (name, aggregator_sources, aggregator, additional_keys)
         )
 
     def gather_results(self, population: pd.DataFrame, event_name: str) -> Dict[str, float]:
-        # Optimization: We store all the producers by pop_filter and groupers
+        # Optimization: We store all the producers by pop_filter and stratifications
         # so that we only have to apply them once each time we compute results.
-        # TODO: uncomment and debug...
-        # for stratification in self._stratifications:
-        #     population = stratification(population)
-        #
-        # for (pop_filter, groupers), observations in self._observations[event_name].items():
-        #     # Results production can be simplified to
-        #     # filter -> groupby -> aggregate in all situations we've seen.
-        #     pop_groups = self.population.query(pop_filter).groupby(list(groupers))
-        #     for measure, aggregator, additional_keys in observers:
-        #         aggregates = pop_groups.apply(aggregator)
-        #         # Keep formatting all in one place.
-        #         yield self._format_results(measure, aggregates, **additional_keys)
-        ...
-
-    def _get_groupers(
+        for stratification in self._stratifications:
+            population = stratification(population)
+
+        for (pop_filter, stratifications), observations in self._observations[
+            event_name
+        ].items():
+            # Results production can be simplified to
+            # filter -> groupby -> aggregate in all situations we've seen.
+            if pop_filter:
+                population = population.query(pop_filter)
+            pop_groups = population.groupby(list(stratifications))
+            for measure, aggregator_sources, aggregator, additional_keys in observations:
+                if aggregator_sources:
+                    aggregates = pop_groups[aggregator_sources].apply(aggregator).fillna(0.0)
```

Comment:
Did you actually see nans in any of your tests? I always want to be careful when we fill nans that we are not masking a bug when we do this.

Reply:
Yes, we see nans if there is no representation for the stratification in the population.